### ISSUE TITLE
cannot determine condition on RVFI with debug external request and interrupt 

### ISSUE THREAD

**Issue by strichmo** [openhwgroup/cv32e40x#204]
### Component
**Component:RTL**: For issues in the RTL (e.g. for files in the rtl directory)

### Steps to Reproduce
1. Use the strichmo fork: https://github.com/strichmo/core-v-verif.git
2. Use this branch: strichmo/pr/common_mk_fw_rule_cleanup
3. Use this bash (Should be head of branch): 2fa98e975aba2fae9660d9d93c3ff23b4c0f2ddd
4. makeuvmt comp_corev-dv gen_corev-dv test TEST=corev_rand_interrupt_debug SEED=1638053781

The instruction retirement at 9451.8 is a debug handler entry due to debug entry request.  However an interrupt is also acknowledge (30) at this instruction retirement.  Therefore the core will save the entry of the next interrupt (0x6954) in DPC.  The current RVFI monitor considers interrupts and debug request in a mutually exclusive manner.  Therefore the RVVI/ISS never sees the interrupts and CSRs start mismatching.

#### Attempted fix
I added code to the RVFI monitor (in UVM) to allow an interrupt request and debug request in the same cycle.  These are signaled to the RVVI in order of interrupt request, debug request, instruction retirement.  This fixes the current condition above.

However when running this fix across other tests, I see the same conditions on RVFI for debug entries where there has not been an interrupt request.

The RVFI signals are (rvfi_valid == 1, rvfi_intr == 1, rvfi_dbg_mode == 0, rvfi_dbg == 3, mcause[31] = 0x1).  Because mcause may be left over from a previous interrupt, the above combination of signals is not a guarantee that we are seeing an interrupt + debug "instruction".  It appears that we need another signal to truly communicate that an interrupt handler was "taken" by the core (even if the actual PC jump is deferred by debug request).

**Comment by Silabs-ArjanB** [comment]
Hi @silabs-halfdan Has this issue been resolved?

**Comment by strichmo** [comment]
@silabs-halfdan Refreshing the testcase with a newer RTL and core-v-verif:

1. On the main core-v-verif repo: 44619257b0d5df818a9e58b5be5ddfcde2884650
2. With Xcelium 21.07.a001:
```
% makeuvmt gen_corev-dv test TEST=corev_rand_interrupt_debug WAVES=1 ADV_DEBUG=1 SEED=393627726 CV_CORE=cv32e40x
```

Error log:
```
UVM_ERROR @ 637578.300 ns : uvme_cv32e40x_core_sb.sv(370) uvm_test_top.env.core_sb [CORESB] CSR Mismatch, order: 36673, csr: dpc, rvfi = 0x00007048, rvvi = 0x00002668
UVM_ERROR @ 637578.300 ns : uvme_cv32e40x_core_sb.sv(370) uvm_test_top.env.core_sb [CORESB] CSR Mismatch, order: 36673, csr: mcause, rvfi = 0x80000012, rvvi = 0x8000001d
UVM_ERROR @ 637578.300 ns : uvme_cv32e40x_core_sb.sv(370) uvm_test_top.env.core_sb [CORESB] CSR Mismatch, order: 36673, csr: mstatus, rvfi = 0x00001880, rvvi = 0x00001888
```

The scenario:
1. As the processor is exitting an interrupt (mret), another interrupt is taken by the processor.  The CSRs are then updated (mcause, mepc).  
2. Before the interrupt handler can be entered, an external debug request is accepted.

**Comment by silabs-oysteink** [comment]
When implementing a fix for issue #196, it seems like we may hit this issue as well. I will revisit that fix once this one is done.
